impeller/renderer/backend/vulkan/fence_waiter_vk.cc

```diff
@@ -47,24 +48,17 @@ class WaitSetEntry {
 FenceWaiterVK::FenceWaiterVK(std::weak_ptr<DeviceHolder> device_holder)
     : device_holder_(std::move(device_holder)) {
   waiter_thread_ = std::make_unique<std::thread>([&]() { Main(); });
-  is_valid_ = true;
 }
 
 FenceWaiterVK::~FenceWaiterVK() {
   Terminate();
-  if (waiter_thread_) {
-    waiter_thread_->join();
-  }
-}
-
-bool FenceWaiterVK::IsValid() const {
-  return is_valid_;
+  waiter_thread_->join();
 }
 
 bool FenceWaiterVK::AddFence(vk::UniqueFence fence,
                              const fml::closure& callback) {
   TRACE_EVENT0("flutter", "FenceWaiterVK::AddFence");
-  if (!IsValid() || !fence || !callback) {
+  if (!fence || !callback) {
```

Comment:
return false if terminate is true?

Reply:
I think we might not actually want to, at least not terminate_. Imagine the following:

1. Get texture/buffer etc
2. Terminate
3. AddFence

In this case, Vulkan still expects the buffer to be live (and there might be other buffers still in the waiter thread), so we should still be queueing the fence. One thing we could do is return if the thread is _actually_ terminated (joined), which isn't the exact same thing?

Reply:
See `StillDestroysFenceIfTerminating` in the unit test suite as well

Reply:
we'll only ever call Terminate or AddFence from the same thread. Once you join, this thread will be blocked on the workload being performed by the fence waiter. So in theory, once Terminate has returned then you'll have already joined the threads - and this should be invalid to add a new fence?

Reply:
So either terminate being true or waiter thread being joined, but make sure you don't let new fences get added and return false or else we're going to have some potentially confusing errors to deal with down the line

Reply:
My mistake. Tweaked the code and unit test.